home_audit.ts

```diff
@@ -0,0 +1,691 @@
+ */
+let _errorMessages = {};
+function getNestedValidationMessages (homeValues) {
+    _errorMessages = {}
+
+    if(!ajv.validate(HesJsonSchema, homeValues)){
+        ajv.errors.forEach((error) => {
+            const {instancePath, params} = error;
+            const errorPath = params.missingProperty ? `${instancePath}/${params.missingProperty}` : instancePath;
+            addErrorMessage(errorPath, convertAJVError(error))
+        });
+    }
+    getCrossValidationMessages(homeValues);
+    return _errorMessages
+}
+
+
+/**
+ * Convert the AJV error into an intelligible error message that the HES system knows how to display
+ * @param {object} errorObj
+ * @return {string|undefined}
+ */
+function convertAJVError(errorObj) {
+    const {keyword, message, schemaPath} = errorObj;
+    const keyArr = schemaPath.split('/');
+    keyArr.shift(); // remove '#'
+
+    // If it's a keyword that's too deep we should pop it to get the right
+    // level for the error message
+    if(['required', 'const'].includes(keyword)) {
+        keyArr.pop();
+    }
+
+    const error_leaf = keyArr.reduce((acc, key) => {
+        return acc[key]
+    }, HesJsonSchema);
+
+    if(error_leaf.error_msg) {
+        return error_leaf.error_msg;
+    }
+
+    switch(keyword) {
+        case 'required':
+            return mandatoryMessage;
+        case 'enum':
+            return `${message}: '${error_leaf.join('\', \'')}'`;
+        case 'if':
+        case 'not':
+        case 'additionalProperties':
+            // Some keywords indicate errors that we don't want to include in our output
+            return undefined;
+        default:
+            return message;
+    }
+}
+
+/**
+ * Get Cross Validation messages for the building to check for other errors.
+ * @param {object} homeValues
+ */
+function getCrossValidationMessages (homeValues) {
+    getAboutObjectCrossValidationMessages(homeValues)
+    getZoneCrossValidationMessages(homeValues.zone, homeValues.about);
+    getSystemCrossValidation(homeValues.systems)
+}
+
+/**
+ * Helper function to add the validation messages easily to the object
+ * @param {string} path Path in the nested schema to the error area in the building object
+ * @param {string} message Validation error message
+ */
+function addErrorMessage(path, message) {
+    if(message) {
+        if (_errorMessages[path] === undefined) {
+            _errorMessages[path] = [];
+        }
+        _errorMessages[path].push(message);
+    }
+}
+
+/**
+ * Cross validations for the "About" object in the nested JSON Schema
+ */
+function getAboutObjectCrossValidationMessages(building) {
+    const {about: {assessment_date: assessmentDate, year_built: yearBuilt}} = building;
+    const today = new Date();
+
+    // The JSON schema ensures that the assessment date is a valid date string, but since it can't validate
+    // minimum or maximum values, we have to do that validation in JavaScript
+    const MIN_ASSESSMENT_DATE = '2010-01-01';
```

Comment:
This is very surprising to me.  Is it just using regex atm them?

Reply:
It's a bit more robust than just a regex:

```
const DATE = /^(\d\d\d\d)-(\d\d)-(\d\d)$/
const DAYS = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

function date(str: string): boolean {
  // full-date from http://tools.ietf.org/html/rfc3339#section-5.6
  const matches: string[] | null = DATE.exec(str)
  if (!matches) return false
  const year: number = +matches[1]
  const month: number = +matches[2]
  const day: number = +matches[3]
  return (
    month >= 1 &&
    month <= 12 &&
    day >= 1 &&
    day <= (month === 2 && isLeapYear(year) ? 29 : DAYS[month])
  )
}
```

...but I don't see a way to define a minimum and maximum, which would be even more complex because the maximum is dynamic. It's possible I just don't know how to define it, though, because a compareDate() function exists and is passed into the validator definition, which suggests it's supposed to be able to handle min/max (else what is it comparing to?)